### PR TITLE
Rust compile time optimizations

### PR DESCRIPTION
**Pull request by Frizi**
### Pull Request Description

Rust compile time optimizations, tracked by https://www.pivotaltracker.com/story/show/183746783

### Important Notes

The majority of time compile optimization in this PR comes from changes in dependency tree. Some dependency duplication got removed, some heavy to compile features got disabled (especially from zip crate). The `prelude` crate is also now compiled only for wasm target. Its (very minimal) usage from our proc-macro crates got removed.

This means the biggest changes will be visible in full rebuild, and significant part of it will also reduce partial builds after modifying anything in prelude or its dependencies.

| | before |  after |
|-|-|-|
|total build time|3m 49.7s|3m 27.6s|
| total compilation units | 909 | 766 |
|ide-view-graph-editor build time| 44.9s | 43.5s |
|ide-view-component-list-panel-grid build time| 30.3s | 28.0s |
|enso_gui.wasm file size (no wasm-opt)| 80.9 MB | 80.3 MB |
|enso_gui.wasm file size (wasm-opt -O3)| 65.4 MB | 65.0 MB |

The frp callstack handling changes should also have positive runtime performance implications when callstacks are enabled. But it's hard to measure that in isolation. It's at least clear that the callstack "sub" method basically disappeared from profiles. Very rougly comparing the opening of component browser (whole event takes ~1s) - before "sub" took 42.4ms, now it takes 0.4ms.

This is only part of the work from the compile time optimization effort - extracted the bits that are most impactful and with safest code changes. A lot is still on the table in the FRP framework core, but for that further work will be needed. So far, the changes there don't make enough difference to consider them for merging.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

### CODE REVIEW

File: lib/rust/ensogl/component/text/src/font/family/src/lib.rs [hunk 1]
```diff
@@ -67,11 +66,24 @@ pub use owned_ttf_parser::Width;
...
+        &self.normalized
+    }
+}
+
+impl std::fmt::Display for Name {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&self.normalized)
+    }
+}
+
```
Comment (kazcw) [inline lib/rust/ensogl/component/text/src/font/family/src/lib.rs]: `prelude` was handling this through `derive_more`. Would it hurt compile time for this crate to use `derive_more` directly?

File: lib/rust/frp/src/data/watch.rs [hunk 2]
```diff
@@ -39,7 +38,7 @@ impl Drop for Handle {
 /// `new_watch` method.
 #[derive(Debug, Clone, CloneRef, Default)]
 pub struct Counter {
-    count: Rc<Cell<usize>>,
+    count: Rc<()>,
```
Comment (kazcw) [inline lib/rust/frp/src/data/watch.rs:41]: Using the implicit count. I love this simplification!

File: lib/rust/generics/src/generic.rs [hunk 3]
```diff
@@ -132,153 +124,154 @@ where T: HasRepr
 // ================
 
 /// Trait for heterogeneous containers like tuples which contain element at index `Ix`.
```
Comment (kazcw) [inline lib/rust/generics/src/generic.rs]: Type variable name `Ix` is outdated in doc comment here and below.

**Review by kazcw — APPROVED** [review]
Nice compile time improvement, and some code cleanup and modernization to boot. Looks great!

File: lib/rust/ensogl/component/text/src/font/family/src/lib.rs (hunk 1 above)
Comment (Frizi) [reply lib/rust/ensogl/component/text/src/font/family/src/lib.rs]: It doesn't impact it. `derive_more` is already compiled before anyway.

File: build/build/Cargo.toml [hunk 4]
```diff
@@ -19,7 +19,7 @@ cfg-if = "1.0.0"
 chrono = { version = "0.4.19", features = ["serde"] }
 clap = { version = "3.1.5", features = ["derive", "env", "wrap_help"] }
 derivative = "2.2.0"
-derive_more = "0.99.17"
+derive_more = "0.99"
```
Comment (farmaazon) [inline build/build/Cargo.toml:22]: Why this change? Is 0.99.17 known to break something? Then maybe we should add more strict constraints?

File: lib/rust/ensogl/component/text/src/font/family/src/lib.rs [hunk 5]
```diff
@@ -44,10 +44,10 @@
 #![warn(unused_import_braces)]
 #![warn(unused_qualifications)]
 
-use enso_prelude::*;
-
+use derive_more::Deref;
+use derive_more::Display;
 use std::collections::HashMap;
-
+use std::fmt::Formatter;
```
Comment (farmaazon) [inline lib/rust/ensogl/component/text/src/font/family/src/lib.rs]: Run `cargo run -p enso-formatter`.

File: lib/rust/frp/src/network.rs [hunk 6]
```diff
@@ -99,25 +103,35 @@ impl Network {
...
     pub fn register<Def: HasOutputStatic>(&self, node: stream::Node<Def>) -> Stream<Output<Def>> {
         let stream = node.clone_ref().into();
-        let node = Box::new(node);
-        self.data.nodes.borrow_mut().push(node);
+        self.register_boxed(Box::new(node));
         stream
     }
 
+    /// Register a boxed node. Non-generic part of registration.
+    #[inline(never)]
```
Comment (farmaazon) [inline lib/rust/frp/src/network.rs:136]: Why do we need this attribute? Add a comment with explanation.

File: lib/rust/frp/src/stream.rs [hunk 7]
```diff
@@ -14,45 +14,53 @@ use crate::data::watch;
 
 /// Owned call stack type.
 #[cfg(feature = "stack-trace")]
-pub type OwnedCallStack = EnabledCallStack;
+pub type OwnedCallStack<'a> = EnabledCallStack<'a>;
```
Comment (farmaazon) [inline lib/rust/frp/src/stream.rs]: Perhaps we should rename this? If it keeps a lifetime, it isn't entirely "Owned", is it?

File: lib/rust/frp/src/stream.rs [hunk 8]
```diff
@@ -14,45 +14,53 @@ use crate::data::watch;
...
 
 
 // === Ops ===
 
 /// Call stack operations available on both enabled and disabled stack implementations.
-pub trait CallStackOps: Default + Display {
+pub trait CallStackOps<'a>: Default + Display {
     /// Create a sub stack trace.
-    fn sub(&self, label: Label) -> Self;
+    fn sub(&'a self, label: Label) -> Self;
```
Comment (farmaazon) [inline lib/rust/frp/src/stream.rs:29]: Can you explain to me why need to specify `'a` here? Is there some implicit that causes returned Self to be also constrained by `'a`?

**Review by farmaazon — APPROVED** [review]
Please address somehow my comments (in the code or in a reply) and then feel free to merge it.

File: build/build/Cargo.toml (hunk 4 above)
Comment (Frizi) [reply build/build/Cargo.toml:22]: We depend on `derive_more` in multiple crates, in some we referred to it with strict version, in others it was just `0.99`. This change ensures that the same version will be used across the board. In this case it isn't that important, but with other crates (like `syn` or `quote`) having the versions specified differently (even if the same is used in the end) can lead to trait resolution errors.

File: lib/rust/frp/src/stream.rs (hunk 7 above)
Comment (Frizi) [reply lib/rust/frp/src/stream.rs]: You can own a struct with a reference :) But I agree, it is a bit inaccurate. I'm going to get rid of this type entirely, as there is no point to refer to that callstack struct as "owned" anyway. If we ever need that, we will need to flatten it into a vec.

File: lib/rust/frp/src/stream.rs (hunk 8 above)
Comment (Frizi) [reply lib/rust/frp/src/stream.rs:29]: It is necessary to tie the input self ref lifetime with the trait, so then the `Self` type in any implementation can store references to that lifetime and let them return from `sub`. Without that tie, you would get a lifetime error in trait implementation.
```
43 |     fn sub(&self, label: Label) -> Self {
   |            - let's call the lifetime of this reference `'1`
44 |         Self { parent: Some(self), label }
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ associated function was supposed to return data with lifetime `'a` but it is returning data with lifetime `'1`
   ```

File: lib/rust/frp/src/network.rs (hunk 6 above)
Comment (Frizi) [reply lib/rust/frp/src/network.rs:136]: Added explanation in the comment.

File: lib/rust/frp/src/stream.rs (hunk 8 above)
Comment (farmaazon) [reply lib/rust/frp/src/stream.rs:29]: Ok, so trait parameter of the trait means that the structure implementing it is constraint by it. It wasn't obvious for me.

Comment (Frizi) [reply lib/rust/frp/src/stream.rs:29]: It doesn't have to be necessarily, but it can. You are free to use that `'a` in the implementation however you want (e.g. in impl bounds), as long as the trait interface is upheld in the end. In this case, the `DisabledCallStack` struct doesn't use it beyond the trait required interface, but `EnabledCallStack` binds its own lifetime arg to the trait, so that the desired `sub` implementation is possible.

File: Cargo.toml [hunk 9]
```diff
@@ -1,4 +1,5 @@
 [workspace]
+resolver = "2"
```
Comment (wdanilo) [inline Cargo.toml:2]: No docs of what it does. Please explain it here with info on how it affects the compile time.

File: app/gui/language/ast/macros/src/lib.rs [hunk 10]
```diff
@@ -110,7 +109,7 @@ fn mk_product_type(
     let ident_nested = format!("{}{}", decl.ident, variant.ident);
     let ident_nested = Ident::new(&ident_nested, Span::call_site());
     let ident = if is_flat { ident_flat } else { ident_nested };
-    let generics = syn::Generics { params, ..default() };
+    let generics = syn::Generics { params, ..Default::default() };
```
Comment (wdanilo) [inline app/gui/language/ast/macros/src/lib.rs:112]: why this change? How does it affect the time?

File: app/gui/language/ast/macros/src/token.rs [hunk 11]
```diff
@@ -28,7 +26,7 @@ pub fn spaceless_ast(input: proc_macro::TokenStream) -> proc_macro::TokenStream
 /// Inner logic for `derive_has_tokens`.
 pub fn derive_for_enum(decl: &syn::DeriveInput, data: &syn::DataEnum) -> TokenStream {
     let ident = &decl.ident;
-    let params = decl.generics.params.iter().collect_vec();
+    let params = decl.generics.params.iter().collect::<Vec<_>>();
```
Comment (wdanilo) [inline app/gui/language/ast/macros/src/token.rs:29]: why this change, how does it affect the time?

File: lib/rust/frp/src/network.rs [hunk 12]
```diff
@@ -99,25 +103,41 @@ impl Network {
...
+    }
+
+    /// Store an arbitrary boxed item.
+    ///
+    /// Force the compiler to never inline this function, so the generated code (including two panic
+    /// handlers) can be shared across all registered items independent of their type.
+    #[inline(never)]
+    fn store_boxed(&self, item: Box<dyn Any>) {
+        self.data.storage.borrow_mut().push(item);
     }
```
Comment (wdanilo) [inline lib/rust/frp/src/network.rs:116]: Does it affect performance of FRP? Did you check it? This will be called a lot of times often in performance heavy places.

File: lib/rust/frp/src/network.rs [hunk 13]
```diff
@@ -99,25 +103,41 @@ impl Network {
...
 
+    /// Register a boxed node. Non-generic part of registration.
+    ///
+    /// Force the compiler to never inline this function, so the generated code (including two panic
+    /// handlers) can be shared across all registered nodes independent of their type.
+    #[inline(never)]
+    fn register_boxed(&self, node: Box<dyn Item>) {
+        self.data.nodes.borrow_mut().push(node);
+    }
```
Comment (wdanilo) [inline lib/rust/frp/src/network.rs:139]: How does it affect the compile time? Does it really make a difference?

File: app/gui/language/ast/macros/src/lib.rs (hunk 10 above)
Comment (Frizi) [reply app/gui/language/ast/macros/src/lib.rs:112]: This is a consequence of removing prelude import. I could also import `default`, but it would require enabling extra nightly feature just for one usage.

File: Cargo.toml (hunk 9 above)
Comment (Frizi) [reply Cargo.toml:2]: The docs for rust resolvers are here https://doc.rust-lang.org/cargo/reference/resolver.html#feature-resolver-version-2

Resolver 2 is the default since edition 2021. Unfortunately, that default doesn't apply to virtual workspaces due to backward compatibility concerns. To me it looks like a missed detail during edition migration.

The potential impact on compile times is that crate feature flags needed in macro crates are treated separately from ones needed in wasm build. Effectively, Less features might be enabled on either side, so less code gets compiled. I don't have exact listing of where this makes a difference right now, but it's good to have it enabled anyway, so future changes have less potential impact.

File: lib/rust/frp/src/network.rs (hunk 12 above)
Comment (Frizi) [reply lib/rust/frp/src/network.rs:116]: I did check those "_boxed" methods together back when I introduced them. It did have some impact, but I don't have the numbers saved. It would now require new separate measurement to quantify how important it is. It might also have an impact on final binary size, so both would need to be checked.

File: lib/rust/frp/src/network.rs (hunk 13 above)
Comment (Frizi) [reply lib/rust/frp/src/network.rs:139]: This is the same story as `store_boxed`, as both were introduced togerher. The registration is the more important of them though, as it is called per each created node, thus the borrow/push needed to be compiled separately for each individual node type in the codebase.

File: app/gui/language/ast/macros/src/token.rs (hunk 11 above)
Comment (Frizi) [reply app/gui/language/ast/macros/src/token.rs:29]: `collect_vec` was part of trait imported from prelude.

File: app/gui/language/ast/macros/src/lib.rs (hunk 10 above)
Comment (wdanilo) [reply app/gui/language/ast/macros/src/lib.rs:112]: Ok, why the prelude was removed here, to make it has less dependencies?

File: lib/rust/frp/src/network.rs (hunk 12 above)
Comment (wdanilo) [reply lib/rust/frp/src/network.rs:116]: Our FRP is not the fastest implementation possible, unfortunately. And thus I care about performance of it as its used in multiple performance critical sections of the renderer. If this affected performance even slightly, I think we should leave it as is and improve FRP with the new implementation in the future.

File: lib/rust/frp/src/network.rs (hunk 13 above)
Comment (wdanilo) [reply lib/rust/frp/src/network.rs:139]: This is not so important to me as the previous ne as this is only when creating nodes, not firing them, so I'm totallyy OK with leaving it this way :)

File: app/gui/language/ast/macros/src/lib.rs (hunk 10 above)
Comment (Frizi) [reply app/gui/language/ast/macros/src/lib.rs:112]: It's a proc-macro crate. Having prelude imported in any of them would result in a lot of crates being compiled twice (once for native target, once for wasm). If we want macro prelude, we should make a separate crate for it. But I don't think it's necessary right now.

File: lib/rust/frp/src/network.rs (hunk 12 above)
Comment (Frizi) [reply lib/rust/frp/src/network.rs:116]: Looking at the code again, this specific function (`store`) is not actually called with massive amount of types, so avoiding monomorphization here probably doesn't have any significant effect. The `register` one is definitely the more important one in that regard. I changed them together because it's the same pattern, so I didn't even considered if it's necessary on both.

Still, granted I don't have any hard data to back it up right now, I don't see any reason why this change would affect perfrormance negatively in any way. Neither `borrow_mut` nor push can be simplified by inlining in our usage patterns, as `borrow_mut` check must always be performed and `Vec` is never preallocated to even have a chance of avoiding the bound check/grow path. If anything, having the code not duplicated might help a little due to better instruction cache/branch predictor utilization. But this is really splitting hairs, as we have much bigger and more fundamental problems with performance, as you already mentioned. In reality it probably doesn't have any noticeable effect and I'd just call this change unnecessary. I'm ok with reverting that tiny portion if you feel it's safer that way.
